### ISSUE TITLE
Lab 04 Keyable Map

### ISSUE THREAD

**Issue by jcng2308** [nus-cs2030/2021-s2#83]
# Summary
I have trouble understanding the errors as I am still unfamiliar with Generics

## Description:
I am getting errors and I do not know how to actually comprehend these errors, please do guide me through!

why get(String) in Student cannot override get(String) in Keyable Map,
V extends Object declared in class KeyableMap

## Code Snippets
  1 import java.util.Map;
  2 import java.util.HashMap;
  3
  4 class KeyableMap<V> implements Keyable {
  5     private final String name;
  6     Map<String, V> map;
  7
  8     public KeyableMap(String name) {
  9         this.name = name;
 10         this.map = new HashMap<>();
 11     }
 12
 13     @SuppressWarnings("unchecked")
 14     public KeyableMap<V> put(V item) {
 15         map.put(item.getKey(), item);
 16         return this;
 17     }
 18
 19     public V get(String name) {
 20         return (V)map.get(name);
 21     }
 22
 23     @Override
 24     public String getKey() {
 25         return this.name;
 26     }
 27
 
`

## Screenshots

Screenshots (if any)
![Error1](https://user-images.githubusercontent.com/64765004/111608586-5adba680-8814-11eb-9c0d-8978d72bbff6.png)
![KeyableMap](https://user-images.githubusercontent.com/64765004/111608599-5e6f2d80-8814-11eb-82db-d83a95f2fc75.png)
![Module](https://user-images.githubusercontent.com/64765004/111608605-5fa05a80-8814-11eb-96a5-39f011c5e956.png)
![Student](https://user-images.githubusercontent.com/64765004/111608621-629b4b00-8814-11eb-8c5d-82858978abeb.png)

**Comment by Keithlimhs** [comment]
Your Module and Student classes need to extend from KeyableMap which includes the generic. So it should be Module extends KeyableMap< Assessment > implements Keyable instead of Module extends KeyableMap implements Keyable

**Comment by Keithlimhs** [comment]
Your KeyableMap class should only be able to take in a class as a generic which implements Keyable as well, so it should be stated. It should be KeyableMap< V extends Keyable > implements Keyable

**Comment by Keithlimhs** [comment]
Also, in your Module and Student constructors, you called the superclass constructor from KeyableMap which alr instantiates a Hashmap as an attribute, hence in the Module and Student constructors you do not need to further declare the attributes this.assessments and this.modules as the super(string) would already instantiate the this.map

**Comment by jcng2308** [comment]
Oh yes thanks keith for reminding me about the super(string) and unnecessary instantiating of map, but i am still having some this issue which was similar to one of the earlier posts, but i cant seem to figure out what it means by Keyable<Assessment> cannot be converted to Module in this context?


![Error](https://user-images.githubusercontent.com/64765004/111621160-9aa98a80-8822-11eb-9f6a-326a17576ad1.png)
![Keyable](https://user-images.githubusercontent.com/64765004/111621170-9e3d1180-8822-11eb-96e5-a95927fc6f83.png)
![Module](https://user-images.githubusercontent.com/64765004/111621178-a1380200-8822-11eb-8b72-ffc3dc2f6034.png)
![Student](https://user-images.githubusercontent.com/64765004/111621184-a301c580-8822-11eb-8e79-c29d7fed56e0.png)

**Comment by Keithlimhs** [comment]
In your KeyableMap constructor, your this.map should specify the HashMap generics, so it should be this.map = new HashMap < String, V >. Also, in your Module and Student classes, you should overwrite the put method so that it returns its own type. So in Module the put method should return Module and in Student, the put method should return Student

**Comment by yuyitan** [comment]
You shouldn't need to override your getKey() and toString() methods in your Module, Student and Roster classes. They don't need to implement Keyable again as they already extend from KeyableMap<V>. However, you will need to override the put() method in Module, Student and Roster to return its own type.

**Comment by jcng2308** [comment]
I just got this error, how do i interpret this "attempting to assign weaker access privileges"

``
@Override
Module put(Assessment assessment) {
   map.put(assessment.getKey(), assessment);
   return this;
}



Module.java:11: error: put(Assessment) in Module cannot override put(V) in KeyableMap
    Module put(Assessment assessment) {
           ^
  attempting to assign weaker access privileges; was public
  where V is a type-variable:
    V extends Keyable declared in class KeyableMap
1 error

**Comment by Keithlimhs** [comment]
Your put method doesnt have to write from scratch. You can just call the super.put method and type cast it to Module

**Comment by jcng2308** [comment]
Thanks so much keith!

**Comment by keiraaa-xrq** [comment]
> Your put method doesnt have to write from scratch. You can just call the super.put method and type cast it to Module

For typecast, do you mean we just write e.g. `(Module) super.put(assessment)`? I got "Exception java.lang.ClassCastException: class REPL.$JShell$15$KeyableMap cannot be cast to class REPL.$JShell$17$Student (REPL.$JShell$15$KeyableMap and REPL.$JShell$17$Student are in unnamed module of loader jdk.jshell.execution.DefaultLoaderDelegate$RemoteClassLoader @16b98e56)" when i tried to do so. May I know how we should typecast properly here since it's like a narrowing conversion if I'm not wrong? My Module extends KeyableMap<Assessment>.

**Comment by Keithlimhs** [comment]
@keiraaa-xrq yup i put it as (Module) super.put(assessment) and it worked fine for me. For your class Module extends KeyableMap, did you specify the generic to extend, for e.g. Module extends KeyableMap< Assessment > ?

**Comment by malcolmsng** [comment]
@keiraaa-xrq i think what you did was in your put method for Student or Module your return type is Keyable<V>, instead of returning the said object i.e. Student or Module respectively.

It should be as follows:
![image](https://user-images.githubusercontent.com/75665163/111893876-fcdfd680-8a40-11eb-900d-d48debb0729c.png)

**Comment by keiraaa-xrq** [comment]
> @keiraaa-xrq yup i put it as (Module) super.put(assessment) and it worked fine for me. For your class Module extends KeyableMap, did you specify the generic to extend, for e.g. Module extends KeyableMap< Assessment > ?



> @keiraaa-xrq i think what you did was in your put method for Student or Module your return type is KeyableMap, instead of returning the said object i.e. Student or Module respectively.
> 
> It should be as follows:
> ![image](https://user-images.githubusercontent.com/75665163/111893876-fcdfd680-8a40-11eb-900d-d48debb0729c.png)

Thanks so much! But I tried to do this but still got ClassCastException in jshell and on codecrunch somehow...

**Comment by keiraaa-xrq** [comment]
@Keithlimhs @malcolmsng Thanks for the help!! It works for me now
